Eric s geo coder

rraub opened the pull request.

Commit:
add basic Reverse Geo Coder php unit and spec test

ESmith92 committed:
Fix whitespace issues

ESmith92 committed:
Add mock of google's geocoder

ESmith92 committed:
Update GoogleReverseGeocoderSpec test

	-add test to check data is returned
	-add test to check "valid key"

ESmith92 committed:
Fix GoogleReverseGeocoderSpec

	-add Search\models namespace
	-give default variables to GeoCoordinate

ESmith92 committed:
Fix whitespace issues

ESmith92 committed:
Fix google-geocoder-mock Coding Standards issues

	-replace "_" with "-" in file name
	-fix whitespace issues
	-fix bracket location issues

ESmith92 committed:
removed "," in google-geocoder-mock

ESmith92 committed:
Update reverse-geocoder-test

 -add test to check Geo_Parser is defined

ESmith92 committed:
fix white space issue in reverse-geocoder-test (unit test)

ESmith92 committed:
add geo-parser-test

	-remove geo-parser is defined test from reverse-geocoder-test
	-add geo-parser is defined test to geo-paarser-test

ESmith92 committed:
add missing return value test

	-add return value test to it_should_return_a_result()

ESmith92 committed:
Update class comment

ESmith92 committed:
Add Geo-Parser spec test

ESmith92 committed:
Update ReverseGeocoder interface

	-add function reverse_geocoding_with_latlin (latitude and longitude)
	-add fucntion reverse_geocoding_with_place_id

ESmith92 committed:
Update google-reverse-geocoder

adds mock functions for calling google's reverse geocoder
	-function to reverse geocode with latitude and longitude
	-function to reverse geocode with place id

ESmith92 committed:
Add geocoder-results-parser

class that will be used to parse the returned values from the geocoder
	-used for both reverse geocoding and geocoding

ESmith92 committed:
Fix GeoParserSpec

	-rename Geo-Parser-Spec file to GeoParserSpec
	-rename class Geo_Parser_Spec to GeoParserSpec

ESmith92 committed:
rename reverse_geocoding_with_latlin function to
reverse_geocoding_with_latlng

	-In
		-google-reverse-geocoder src/providers
		-reverse-gecoder - src/service

ESmith92 committed:
Move google-geocoder-mock

	-from: test\spec\support 
	-to: test\spec\support\Mocks

ESmith92 committed:
Update GoogleReverseGeoCoder for the new mock location

ESmith92 committed:
Fix GoogleReverseGeocoderSpec and GeoParserSpec

	-change provider spec test directory: match spec structure
	-change support spec test directory: match spec structure
	-change spec namespaces to fit new directories
	-change Geo_Parser class name to GeoParser to match spec test name
	-change it_is_initializable in GeoParserSpec to look for type GeoParser

ESmith92 committed:
fix GeoPaserSpec test it_should_return_a_result

	-replace ->shouldBeCalled to ->shouldBeString()

ESmith92 committed:
Remove google-reverse-geocoder func: place id

	-place id is specific to googles geocoding

ESmith92 committed:
Add google-geocoder-mock to phpspec and phpunit

	-add to be autoinjected when testing with mock

ESmith92 committed:
Fix white space issue

	-add newline to end of file

ESmith92 committed:
fix geo-parser-test.php

	-change class defined asserto 
		from: SearchApi\Support\Geo_Parser
		to: SearchApi\Support\GeoParser

ESmith92 committed:
Fix geocoder to pass spec tests

	-Update class Google_Geocoder_Mock
		-add test to check if $latlin is null
		
	-update class GoogleReverseGeocoder
		-Add class var $api_key
		-Add Constructor to init $api_key
		-Fix get_locations
			-add instance of google geocoder mock
			-change to instance of mock (instead of static)
			-add check if key is valid before parser
		
	-Update GoogleReverseGeocoderSpec
		-update it_should_return_a_result
			-add $api_key handler
		-update it_should_return_invalid_key
			-add $api_key handler

kenprice committed:
Merge branch 'develop' into EricS-GeoCoder

rraub commented on test/spec/SearchApi/Support/GeoParserSpec.php:
```diff
@@ -0,0 +1,22 @@
+  }
+
+  function it_should_return_a_result() {
+  	$this->reverse_geocoder_parser( 'test' )->shouldBeString();
```
Your indentation isn't consistent here, I'll have to look into why the coding standards didn't catch this.

ESmith92 replied:
I checked my code and it appears to have the correct indenting in code but git shows it over farther

idmontie replied:
Is it being committed with an actual tab? You should set your IDE  to convert tabs to spaces, not just have them show up as 2 spaces.

ESmith92 replied:
it should be set to 2 spaces, ill double check though

rraub commented on src/providers/google-reverse-geocoder.php:
```diff
@@ -2,11 +2,30 @@
+  	$this->api_key = $key;
+  }
+
+  public function get_locations( Models\GeoCoordinate $geo_coordinate ) {
```
Could you add some basic comments to this function?

rraub commented on src/providers/google-reverse-geocoder.php:
```diff
@@ -2,11 +2,30 @@
+  	$mock = new Mocks\Google_Geocoder_Mock();
+  	$geocoding_results = $mock->reverse_geocoding( $geo_coordinate, $this->api_key );
+  	if ( $geocoding_results === 'Invalid Key' ) {
+  	  return 'Invalid Key';
```
Returning a string as an error state might not be the easiest thing for who ever is using this to catch. Especially when it looks like your actual data being returned is also be a string.

idmontie replied:
It would be better to just throw an Exception with the message "Invalid Key"; because, well, its an exceptional circumstance.

ESmith92 committed:
Change GoogleReverseGeocoderSpec

	-add null test for api_key
		-found out the key is optional
		-have test if going to use key
	-change invalid key test
		-check for error thrown

ESmith92 committed:
change geo_parser to be struct injected in GoogleReverseGeocoder

	-fix GoogleReverseGeocoder
		-add in parser to struct
		-fix parser method call
		
	-fix GoogleReverseGeocoderSpec
		-update it_should_throw_invalid_key & it_should_return_a_result to
handle struct injection

ESmith92 committed:
fix whitespace issues

ESmith92 committed:
Add it_should_return_a_result_without_key spec test to
GoogleReverseGeocoder

ESmith92 committed:
Add invalid key exception to GeoParser

	-GoogleReverseGeocoder
		-temporarily check if key is invalid
		-change "service" result to error message
		
	-GeoParser
		-check for the error message
			-throw error if error message found
			
	-Google_Geocoder_Mock
		-hold default error message
			-model on google geocoding error message

ESmith92 committed:
Remove need for mock of google geocoding service

	-GoogleReverseGeocoder
		-remove call to mock
		-replace with temp result string
		
	-GeoParser
		-add check that result is utf8 encoded
		-add basic json decoder
		-add check for invalid key by status
			-check if 'status' equals 'REQUEST_DENIED'

ESmith92 committed:
add invalid key test to GeoParserSpec

ESmith92 committed:
add test to check for exception if json is invalid

ESmith92 committed:
add code to pass bad json exception test

-geocoder-results-parser
	-add if statement to check for bad json
	-update if statement for invalid key
		-check if status key is avaliable
		
-google-reverse-geocoder
	-change dfault input to valid json input

ESmith92 committed:
update geocoder to call google's geocoder service

-GoogleReverseGeocoder
	-formatts the service url
		-checks if key is given
		-can add more control if needed
	-create curl
		-create curl
		-set up option
		-execute curl (service call)
		-check if curl went through
	-send results to parser
	
-GoogleReverseGeocoderSpec
	-update lat and lng to correct location

ESmith92 committed:
add code to parse results of google geocoder

-GoogleReverseGeocoderSpec
	-update test to check for a return of an array

-GeoParserSpec
	-update test to check for a return of an array
	
-GeoParser
	-add code to loop through and parse out address components of google
geocoder
		-add address components to array of search terms
		-returns the array

ESmith92 committed:
fix whitespace issues

ESmith92 committed:
remove geocoder mock

ESmith92 committed:
Merge branch 'develop_update' into EricS-GeoCoder

ESmith92 commented:
alright this branch is ready to be checked for a merge

rraub commented on src/support/geocoder-results-parser.php:
```diff
@@ -0,0 +1,76 @@
+ * Class Geo_Parser - Parses the returned result from a Geocoder
+ */
+class GeoParser {
+  public function reverse_geocoder_parser( $json_results ) {
```
This function is doing too much, you can slim it up by separating out the creation of the $search_term_array into another function and have reverse_geocoder_parser call that. This will make testing easier too since you can test the logic for creating the $search_term_array separately from the logic that encodes and decodes the json response.

ESmith92 replied:
Alright, I'll try to do this tonight

rraub commented on test/spec/SearchApi/Support/GeoParserSpec.php:
```diff
@@ -0,0 +1,32 @@
+  }
+
+  function it_should_return_an_empty_array() {
+  	$this->reverse_geocoder_parser( '{"test":"test"}' )->shouldBeArray();
```
This doesn't actually test if the result is an empty array.

rraub commented on test/spec/SearchApi/Providers/GoogleReverseGeocoderSpec.php:
```diff
@@ -0,0 +1,45 @@
+    $this->shouldHaveType( 'SearchApi\Providers\GoogleReverseGeocoder' );
+  }
+
+  function it_should_have_key_null() {
```
maybe a better name is `it_shouldnt_have_an_api_key_by_default`

ESmith92 replied:
It was a test to make sure it is null by default.
How about it_should_have_null_for_api_key

rraub replied:
Your testing the default case, which is why I would like to see the word 'default' in the test name. Otherwise it might sound more like you are just testing how the class works with a null api_key.

ESmith92 replied:
Ah, I see what you mean

ESmith92 replied:
Oh I just reread your suggested name, I missed the "shouldn't" and read it as "should"

rraub commented on src/providers/google-reverse-geocoder.php:
```diff
@@ -2,11 +2,58 @@
+  public function __construct( $key = null, $parser = null ) {
+  	$this->api_key = $key;
+
+  	if ( $parser === null ) {
```
The indentation is off here too, it has a mix of tabs and spaces

ESmith92 replied:
Ill check if there a way to tell the ide i am using to use space when I hit enter... I think that is the issue happening with these tabs.

ESmith92 committed:
fix whitespace issue

ESmith92 committed:
rename GoogleReversGeocoderSpec test it_should_have_key_null

	-rename to: it_shouldnt_have_an_api_key_by_default

ESmith92 committed:
fix GeoParserSpec it_should_return_an_empty_array

	-add code to check that the array is empty
		-not just check that it is an array

ESmith92 committed:
split creation of SearchTerm array from jdon_decoding in GeoParser

-GeoParser
	-rename reverse_geocoder_parser to reverse_geocoder_json_decoder
	-add function reverse_geocoder_pareser
	-move SearchTerms array from reverse_geocoder_json_decoder to
reverse_geocoder_parser

-GoogleReverseGeocoder
	-updated get_locations to handle parser being split into 2 functions
	
-GeoParserSpec
	-update test to handle parser being split into 2 functions

ESmith92 committed:
update GeoparserSpec

-add new tests
	-it_should_return_an_empty_array_when_passed_null
	-it_should_return_an_empty_array_when_passed_empty_array
	-it_should_return_an_empty_array_when_passed_bad_formatted_array
	
-rename test: it_should_return_an_empty_array
	-to: it_should_return_an_empty_array_after_json_decoding_and_parsing

ESmith92 committed:
fix GeoParser function reverse_geocoder_parser to handle null case

ESmith92 committed:
fix travis-ci issues

-GeoParserSpec
	-fix double quotes needing to be single quotes
	-removed newline whitespace
	
-GeoParser
	-retype comment to show that it is not commented code

ESmith92 commented:
this is ready to be check again

idmontie commented:
:+1: Looks good. I'll let @rraub check it in, but he is out today (Wednesday).

rraub commented on test/spec/SearchApi/Support/GeoParserSpec.php:
```diff
@@ -0,0 +1,59 @@
+  	$result->shouldBeArray();
+  	$result->shouldHaveCount( 0 );
+  }
+  // end of reverse_geocoder_parser tests
```
Where is the test case for the reverse_geocoder_parser returning a real value? Not an empty array.

ESmith92 committed:
add test to GeoParserSpec

-add test to check for real json object parseing
	-add test json string
	-test return is array and has 2 elements
	-test values of the elements

rraub commented:
:+1:
